app/models/bulk_download.rb

```diff
@@ -258,10 +258,82 @@ def bulk_download_ecs_task_command
     return nil
   end
 
+  def generate_download_file
+    samples = Sample.where(id: pipeline_runs.pluck(:sample_id))
+    projects = Project.where(id: samples.pluck(:project_id))
+
+    # Compute cleaned project name once instead of once per sample.
+    cleaned_project_names = {}
+    projects.each do |project|
+      cleaned_project_names[project.id] = project.cleaned_project_name
+    end
+
+    failed_sample_ids = []
+
+    if download_type == BulkDownloadTypesHelper::SAMPLE_TAXON_REPORT_BULK_DOWNLOAD_TYPE
+      s3_tar_writer = S3TarWriter.new(download_output_url)
+      s3_tar_writer.start_streaming
+
+      pipeline_runs.includes(sample: [:project]).map do |pipeline_run|
+        begin
+          tax_details = ReportHelper.taxonomy_details(pipeline_run.id, get_param_value("background"), TaxonScoringModel::DEFAULT_MODEL_NAME, ReportHelper::DEFAULT_SORT_PARAM)
+          report_csv = ReportHelper.generate_report_csv(tax_details)
+          s3_tar_writer.add_file_with_data(
+            "#{get_output_file_prefix(pipeline_run.sample, cleaned_project_names)}" \
+              "taxon_report.csv",
+            report_csv
+          )
+        rescue
+          failed_sample_ids << pipeline_run.sample.id
+        end
+      end
+
+      s3_tar_writer.close
```

Comment:
do you think it's possible that the `close` actions will hang? / should we have a timeout for this?

Reply:
That's a good point.  I think it's more likely that the `process_status` method will hang because that's the one that waits for the `aws s3 cp` thread to complete. Any suggestions for the best way to implement a timeout?

Reply:
I think this looks like what we want? https://flushentitypacket.github.io/ruby/2015/02/21/ruby-timeout-how-does-it-even-work.html

Reply:
There's a link at the bottom of that that says ruby Timeout is "dangerous". I'm also not sure what a good timeout to set is. Maybe a monitoring solution like we have for our long-running samples would be best.

Would you be okay with us deferring this until later? I can add some log statements to this Resque task which will hopefully make it more obvious if something has hung.

Reply:
yep sounds good 👍